Sources/Networking/Core/Upload/UploadAPIManager.swift

```diff
@@ -0,0 +1,293 @@
+        retryConfiguration: RetryConfiguration?,
+        data: Data?,
+        response: URLResponse?,
+        error: Error?
+    ) {
+        Task {
+            guard let uploadTask = await uploadTasks.getValue(for: endpointRequest.id) else {
+                return
+            }
+
+            var state = UploadTask.State(task: uploadTask.task)
+            if let data, let response {
+                state.response = try await responseProcessors.process(
+                    (data, response),
+                    with: urlRequest,
+                    for: endpointRequest
+                )
+
+                try await uploadTask.complete(with: state)
+
+                // Cleanup on successful task completion
+                await uploadTask.resetRetryCounter()
+                await uploadTasks.set(value: nil, for: endpointRequest.id)
+            } else if let error {
+                do {
+                    // URLError.Code.cancelled is thrown if the URLSessionTask is cancelled.
+                    // Consider this action intentional, thus the request won't be retried.
+                    guard !state.cancelled else {
+                        throw error
+                    }
+
+                    try await uploadTask.sleepIfRetry(
+                        for: error,
+                        retryConfiguration: retryConfiguration
+                    )
+
+                    try await self.uploadRequest(
+                        uploadTask.uploadable,
+                        request: uploadTask.endpointRequest,
+                        retryConfiguration: retryConfiguration
+                    )
+                } catch {
+                    state.error = await errorProcessors.process(
+                        error,
+                        for: uploadTask.endpointRequest
+                    )
+
+                    // No cleanup in case the task will be retried.
+                    try await uploadTask.complete(with: state)
+                }
+            }
+        }
+    }
+
+    /// The implementation uses completion closure version of the upload task instaed of the async versions.
+    ///
+    /// The async versions could be used, however, if we go down that route, we'll need to make some slight
+    /// changes to the implementation, because:
+    /// - The async versions don't return `URLSessionTask` and will be suspended, which means returning immediately
+    ///   as it is now wouldn't be possible. So we'll need to consider what to return to the client, if anything at all.
+    /// - We'll need to handle errors and responses from the request using delegates.
+    func sessionUploadTask(
+        with uploadable: Uploadable,
+        for request: URLRequest,
+        completionHandler: @escaping @Sendable (Data?, URLResponse?, Error?) -> Void
+    ) -> URLSessionUploadTask {
+        switch uploadable {
+        case let .data(data):
+            return urlSession.uploadTask(
```

Comment:
Thread 12: "Completion handler blocks are not supported in background sessions. Use a delegate instead."

Due to the use of completion blocks, UploadAPIManager cannot be used in background!!

Reply:
https://github.com/strvcom/ios-networking/pull/48